presto-iceberg/src/test/java/com/facebook/presto/iceberg/IcebergDistributedTestBase.java

```diff
@@ -859,68 +868,168 @@ public void testTableWithPositionDelete(String fileFormat)
+        Table icebergTable = updateTable(tableName);
+
+        List<Long> partitions = Arrays.asList(1L, 2L, 3L, 17L, 24L);
+        for (long nationKey : partitions) {
+            writeEqualityDeleteToNationTable(icebergTable, ImmutableMap.of("regionkey", 1L), ImmutableMap.of("nationkey", nationKey));
+        }
+        testCheckDeleteFiles(icebergTable, partitions.size(), partitions.stream().map(i -> EQUALITY_DELETES).collect(Collectors.toList()));
+        assertQuery(session, "SELECT * FROM " + tableName, "SELECT * FROM nation WHERE regionkey != 1");
+        assertQuery(session, "SELECT nationkey FROM " + tableName, "SELECT nationkey FROM nation WHERE regionkey != 1");
+        assertQuery(session, "SELECT name FROM " + tableName, "SELECT name FROM nation WHERE regionkey != 1");
+    }
+
+    @Test(dataProvider = "equalityDeleteOptions")
+    public void testEqualityDeletesWithHiddenPartitions(String fileFormat, boolean pushdownEnabled)
+            throws Exception
+    {
+        Session session = deleteAsJoinEnabled(pushdownEnabled);
+        String tableName = "test_v2_row_delete_" + randomTableSuffix();
+        assertUpdate("CREATE TABLE " + tableName + " with (format = '" + fileFormat + "', partitioning = ARRAY['bucket(nationkey,100)']) AS SELECT * FROM tpch.tiny.nation order by nationkey", 25);
+        Table icebergTable = updateTable(tableName);
+
+        PartitionTransforms.ColumnTransform columnTransform = PartitionTransforms.getColumnTransform(icebergTable.spec().fields().get(0), BigintType.BIGINT);
+        BlockBuilder builder = BigintType.BIGINT.createFixedSizeBlockBuilder(5);
+        List<Long> partitions = Arrays.asList(1L, 2L, 3L, 17L, 24L);
+        partitions.forEach(p -> BigintType.BIGINT.writeLong(builder, p));
+        Block partitionsBlock = columnTransform.getTransform().apply(builder.build());
+        for (int i = 0; i < partitionsBlock.getPositionCount(); i++) {
+            writeEqualityDeleteToNationTable(icebergTable, ImmutableMap.of("regionkey", 1L), ImmutableMap.of("nationkey_bucket", partitionsBlock.getInt(i)));
+            String updatedPartitions = partitions.stream().limit(i + 1).map(Object::toString).collect(Collectors.joining(",", "(", ")"));
+            assertQuery(session, "SELECT * FROM " + tableName, "SELECT * FROM nation WHERE NOT(regionkey = 1 AND nationkey IN " + updatedPartitions + ")");
+        }
+        testCheckDeleteFiles(icebergTable, partitions.size(), partitions.stream().map(i -> EQUALITY_DELETES).collect(Collectors.toList()));
+        assertQuery(session, "SELECT * FROM " + tableName, "SELECT * FROM nation WHERE regionkey != 1");
+        assertQuery(session, "SELECT nationkey FROM " + tableName, "SELECT nationkey FROM nation WHERE regionkey != 1");
+        assertQuery(session, "SELECT name FROM " + tableName, "SELECT name FROM nation WHERE regionkey != 1");
+    }
+
```

Comment:
```suggestion

    @Test(dataProvider = "equalityDeleteOptions")
    public void testEqualityDeletesWithHiddenPartitionsEvolution()
            throws Exception
    {
        Session session = deleteAsJoinEnabled(true);
        String tableName = "test_v2_row_delete_" + randomTableSuffix();
        assertUpdate("CREATE TABLE " + tableName + "(a int, b varchar) with (format = 'parquet')");
        assertUpdate("insert into " + tableName + " values (1, '1001'), (2, '1002'), (3, '1003')", 3);
        Table icebergTable = updateTable(tableName);
        writeEqualityDeleteToNationTable(icebergTable, ImmutableMap.of("a", 2));
        assertQuery(session, "SELECT * FROM " + tableName, "values(1, '1001'), (3, '1003')");

        assertUpdate("alter TABLE " + tableName + " add column c int with(partitioning = 'bucket(2)')");
        assertUpdate("insert into " + tableName + " values (6, '1004', 1), (6, '1006', 2)", 2);
        icebergTable = updateTable(tableName);
        PartitionTransforms.ColumnTransform columnTransform = PartitionTransforms.getColumnTransform(icebergTable.spec().fields().get(0), INTEGER);
        BlockBuilder builder = BIGINT.createFixedSizeBlockBuilder(1);
        List<Integer> partitions = Arrays.asList(1, 2);
        partitions.forEach(p -> BIGINT.writeLong(builder, p));
        Block partitionsBlock = columnTransform.getTransform().apply(builder.build());
        writeEqualityDeleteToNationTable(icebergTable, ImmutableMap.of("a", 6, "c", 2),
                ImmutableMap.of("c_bucket", partitionsBlock.getInt(0)));
        assertQuery(session, "SELECT * FROM " + tableName, "values(1, '1001', NULL), (3, '1003', NULL), (6, '1004', 1)");

        assertUpdate("alter TABLE " + tableName + " add column d varchar with(partitioning = 'truncate(2)')");
        assertUpdate("insert into " + tableName + " values (6, '1004', 1, 'th001'), (6, '1006', 2, 'th002')", 2);
        icebergTable = updateTable(tableName);

        writeEqualityDeleteToNationTable(icebergTable, ImmutableMap.of("a", 6, "c", 1),
                ImmutableMap.of("c_bucket", partitionsBlock.getInt(0), "d_trunc", "th"));
        testCheckDeleteFiles(icebergTable, 3, ImmutableList.of(EQUALITY_DELETES, EQUALITY_DELETES, EQUALITY_DELETES));
        assertQuery(session, "SELECT * FROM " + tableName, "values(1, '1001', NULL, NULL), (3, '1003', NULL, NULL), (6, '1004', 1, NULL), (6, '1006', 2, 'th002')");
    }
```

Seems in this test case the record `(6, '1004', 1, NULL)` would be wrongly deleted. The reason might be that, when we try to scan files for equality delete table of specific delete schema  by IcebergUtil.getDeleteFiles(...), we should take the partition columns of delete schema as well. Or different delete schema with same equality delete columns and different partition columns would lead in some problems.

Reply:
Great catch! thanks for writing a test to reproduce this! I pushed a fix for this.